Home: Add model description text on web UI

dongbohu opened the pull request:
This PR is to fix issue #275. I also fixed a couple of minor bugs that I found when working on this one.

On the backend:
(1) Add a new `desc_html` field in `MlModel`;
(2) A new migration file was generated (no need to review); 
(2) Management command and deployment is also changed.

On the front end:
(1) The corresponding `Home` js and html template files are revised;
(2) The `MlModelTracker` service is modified because of the new field in the model;
(3) A minor issue in `ml-model-selector` component is corrected.

dongbohu commented:
Codeclimate reported a few issues. I keep them on purpose because it is probably better to have the whole URL on one line.

mhuyck commented:
When did Codeclimate start complaining about things like this?

> Function add_ml_model has 5 arguments (exceeds 4 allowed). Consider refactoring.

That seems new to me, and I don't think we have anything like this included in our coding style guidelines so maybe we should see why it's suddenly popping up. Please correct me if I am wrong.

mhuyck commented on adage/analyze/models.py:
```diff
@@ -149,6 +149,7 @@ class MLModel(models.Model):
     # False (not directed).
     directed_g2g_edge = models.BooleanField(default=False)
     g2g_edge_cutoff = models.FloatField(default=0.0)
+    desc_html = models.CharField(max_length=2048, blank=True)
```
I was a little surprised to find that Django doesn't have any sort of `HTMLField` but I guess a `CharField` can hold HTML just fine. Did you consider using a `TextField` here instead? I don't think I have a strong opinion one way or another. The main difference is that `TextField` has no pre-set limit on length.

There is great potential for abuse of unfiltered user-provided HTML on web sites, but in this case we only allow an administrator to supply a `desc_html` (via the management command) so as long as that remains true we should be safe on that front.

dongbohu replied:
I thought about it when adding the new field. But I saw some online article claiming that `CharField` is slightly faster than a `TextField` (I forgot from where though). So I eventually used the former.

mhuyck replied:
I'd guess the trade-off is that `CharField` is less efficient in storage or memory space, but the defined `max_length` makes it faster to execute.

mhuyck commented on fabfile/adage_server.py:
```diff
@@ -197,6 +197,17 @@ def import_data_and_index():
+    'been shared is available and has been described in '
+    '<a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-017-1905-4">'
+    'Tan et al., 2017. <em>BMC Bioinformatics.</em></a>'
+
```
Indentation seems a little off here so I'm wondering why Codeclimate didn't pick that up. Does this pass muster with a Python linter?

dongbohu replied:
Yes, pep8 is happy with it.

mhuyck commented on interface/src/app/home/home.tpl.html:
```diff
@@ -20,32 +20,36 @@ <h3>ADAGE: Home</h3>
+      <a ui-sref="analyze({mlmodel: modelInfo.id})">
+        Explore assays and experiments
+      </a>
+    </li>
```
Nice! Thank you for cleaning up the HTML here.

mhuyck commented on interface/src/app/mlmodel/mlModelComponents.js:
```diff
@@ -6,7 +6,7 @@ angular.module('adage.mlmodel.components', [
   templateUrl: 'mlmodel/selector.tpl.html',
   bindings: {
-    selectedMlModel: '=',
+    selectedMlModel: '<',
```
👍

mhuyck commented on interface/src/app/mlmodel/mlModelComponents.js:
```diff
@@ -65,7 +65,7 @@ angular.module('adage.mlmodel.components', [
         },
         function error(err) {
-          MlModelTracker.reset();
+          MlModelTracker.init();
```
I was wondering if it was safe to call `.init()` in an error handler, so I looked up `MLModelTracker` (should really be defined in `src/app/mlmodel`, by the way, because I had a hard time finding it in `utils.js`) and found that it is indeed safe because all it does is clear its properties. I think the name `reset` or `clear` would be more accurate, but I understand why you're using `init()`. Perhaps a simple way around this would be to actually define a `reset()` and call that from `init()`.

Anyway, it's not a huge concern since `init()` currently doesn't currently do anything unsafe, but it would become a problem down the road if we added some code there that wouldn't be good to call during an error handler. I'd recommend at minimum a comment in `MLModelTracker.init()` to guard against that.

dongbohu replied:
Good point. Part of the reason why `MlModelTracker` is located in `utils.js` is because this service is needed by almost every front-end module, and what it does is keep track of the current mlmodel, and it is not directly related to any front-end display.
 
The distinction of `init()` and `reset()` makes a lot of sense. Let me add a `reset` method which will call `init()` for now.

mhuyck replied:
Why don't we *define* the `MLModelTracker` service in `/src/app/mlmodel`, so it can be with all other `MLModel` code, and (since, as you say, it's needed everywhere) *include* it in `utils.js`?

mhuyck replied:
To be more precise: when I say "include", I mean use Angular's dependency mechanism to ensure it is available.

dongbohu replied:
Right now `mlmodel` directory has two js files: `mlModelComponents.js` and `resource.js`. Neither of them seems to be a good place for `MLModelTracker` service. `MlModelTracker` doesn't rely on any external modules, I would argue that it is more like a utility. (But I agree that its name may be misleading.) I also don't want to make this PR too complex.

mhuyck replied:
I agree that this is not an essential part of this PR... it just came up because I had trouble tracking down `MLModelTracker` during the review since it wasn't where I expected to find it. Since we have more details to discuss let's file an issue so it doesn't hold up the main point of this PR.

dongbohu replied:
Done!

mhuyck commented on interface/src/app/home/home.js:
```diff
@@ -2,6 +2,7 @@
  * adage.home module.
  */
 angular.module('adage.home', [
+  'ngSanitize',
```
It looks as if `ngSanitize` is not used anywhere. Am I missing something?

dongbohu replied:
According to:
https://docs.angularjs.org/api/ngSanitize/service/$sanitize
`ng-bind-html` seems to automatically call `$sanitize`, which requires `ngSanitize`.

dongbohu replied:
I tested the example at the end of this page:
https://docs.angularjs.org/api/ng/directive/ngBindHtml
If `ngSanitize` module is removed from dependency, the page won't be rendered correctly. Here is the error:
```
Error: [$sce:unsafe] http://errors.angularjs.org/1.6.9-build.5530+sha.07d84dd/$sce/unsafe
    at https://code.angularjs.org/snapshot/angular.min.js:7:93
    at f (https://code.angularjs.org/snapshot/angular.min.js:157:225)
    at getTrusted (https://code.angularjs.org/snapshot/angular.min.js:158:462)
    at Object.c.(anonymous function) [as getTrustedHtml] (https://code.angularjs.org/snapshot/angular.min.js:160:175)
    at https://code.angularjs.org/snapshot/angular.min.js:286:459
    at m.$digest (https://code.angularjs.org/snapshot/angular.min.js:148:416)
    at m.$apply (https://code.angularjs.org/snapshot/angular.min.js:151:505)
    at https://code.angularjs.org/snapshot/angular.min.js:22:117
    at Object.invoke (https://code.angularjs.org/snapshot/angular.min.js:44:442)
    at c (https://code.angularjs.org/snapshot/angular.min.js:22:38)
```

mhuyck replied:
👍 

Since that connection is not obvious, can we put a short comment at line 5 saying something like:

```javascript
  'ngSanitize',  // used by ng-bind-html
```

dongbohu commented:
@mhuyck: I believe I have addressed all your concerns except the relocation of `MlModelTracker` service. Please let me know if you have any other questions. Thanks.

mhuyck commented on interface/src/app/utils.js:
```diff
@@ -34,9 +35,12 @@ angular.module('adage.utils', [])
   };
 
+  this.reset = this.init;
+
```
Hah! Clever.

mhuyck commented:
Ok, this works for me. Thank you, @dongbohu!
